src/apps/ontology/sc/native_builder.py

```diff
@@ -0,0 +1,118 @@
+from trezor import wire
+
+from . import builder, opcode
+from .. import writer
+
+
+class ParamStruct:
+    """
+    Special struct for smart contract argument passing
+    """
+
+    def __init__(self, arr: list):
+        self.arr = arr
+
+
+def build_native_call(funcName: str, params: list, contract: str) -> bytes:
+    """
+    Builds native contract call
+    """
+    ret = bytearray()
+
+    _write_native_code_script(ret, params)
+    builder.write_push_bytes(ret, bytes(funcName, "ascii"))
+    builder.write_push_hexstr(ret, contract)
+    builder.write_push_int(ret, 0)
+    writer.write_byte(ret, opcode.SYSCALL)
+    builder.write_push_bytes(ret, b"Ontology.Native.Invoke")
+
+    return ret
+
+
+def _write_native_code_script(ret: bytearray, arr: list) -> None:
```

Comment:
this seems more complicated than it needs to be

the usage in your code is for exactly two cases:
1. `params` is a list of a single ParamStruct, and
2. `params` is a list of a list of a single ParamStruct

also, in all cases the ParamStruct seems to not be recursive, i.e., it is always a list of simple values

I propose we convert this code so that it _always_ takes a list of simple values, and an additional argument which controls whether the input should be considered an array of n elements
so let's say:
```python
def _write_native_code_script(ret: bytearray, params: list, array_len: int = None):
    write_byte(NEWSTRUCT)
    (...)
    for val in params:
        if isinstance(...):
            (...)
        (...)
    
    write_byte(FROMALTSTACK)
    if array_len is not None:
        write_push_int(array_len)
```

This won't be a generic script builder anymore, but we don't seem to need that, and it makes the code simpler to reason about. It will also let us get rid of `ParamStruct` and pass the lists directly.

Or is there a reason to keep the code generic, that I'm not seeing?